### ISSUE TITLE
Pytorch 1.7/Cuda 11.1 binaries

### ISSUE THREAD

**Issue by LukeAI** [pytorch/pytorch#47109]
Will Pytorch 1.7/Cuda 11.1 binaries be provided or only CUDA 11.0?

**Comment by ptrblck** [comment]
PyTorch 1.7.0 binaries will ship with CUDA11.0, as we were unable to build them with CUDA11.1 due to an internal error in `nvprune`.

**Comment by LukeAI** [comment]
ok, thanks for reply! Is it likely that this will be resolved and CUDA 11.1 binaries will be provided in the near-future?

**Comment by seemethere** [comment]
We are currently planning for a 1.7.1 although a release date remains TBD, this update will most likely ship with an updated version of CUDA 11 (hopefully CUDA 11.2 at that point)

**Comment by LukeAI** [comment]
I notice that binaries are provided for CUDA 9.2, 10.1, 10.2 but NOT 10.0   - should I expect support for 11.0 to be dropped in the near-future too? would 10.2 be a better choice if I want to choose a stable long term support standard?

**Comment by stas00** [comment]
Your choice will also depend on your hardware, @LukeAI - if you plan on using rtx30, you need at least cuda-11.0

**Comment by LukeAI** [comment]
@stas00 yeah - I would really like the option to use rtx3080s in the future without having to upgrade everything, but if support for 11.0 gets dropped quickly (as it was with 10.0) then I might be better off going with 10.2 and just moving all of the machines to 11.X when I actually get the cards - but any cuda upgrade across many systems is a big job so I feel slightly in limbo here.

**Comment by stas00** [comment]
I'm not sure why you find that the upgrade is cumbersome, since it's all bundled together and not touching on the system files it should be very easy to move from cudatoolkit-10.2 to cudatoolkit-11.x (assuming the code is not impacted), or 11.0 to 11.1
```
conda install pytorch torchvision torchaudio cudatoolkit=11.0 -c pytorch
```

it's not that the support for lower minor versions is dropped, it's just that binaries are not being built anymore for those (.0 typically), since higher minor versions have bugs fixed in them, so there is no point providing a binary for a buggy version when there is a bug-free version.

**Comment by LukeAI** [comment]
@stas00  unfortunately using conda to manage our cuda is not really an option since we use ROS and conda has all sorts of conflicts with that, so we do system installs of CUDA as per [nvidia's](https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html) instructions.

**Comment by stas00** [comment]
Do you have the same issue with `pip`, since you could do:
```
pip install torch==1.7.0+cu110 torchvision==0.8.1+cu110 torchaudio===0.7.0 -f https://download.pytorch.org/whl/torch_stable.html
```

but, yes, if you have to build from source against the system-wide cuda that would be a pain. In such case I'd build a binary package for whatever package manager your system uses and then it'd be easy to replicate across machines.

> we use ROS

Do you mean Robot Operating System (ROS)?

**Comment by LukeAI** [comment]
That pip install does appear to work, although I can't properly test it right now as the nvidia driver on my current machine is too old for CUDA11. we don't just use CUDA for pytorch so not going to be a complete solution, but good to know that's an option, thankyou!
yes, I do mean the Robot Operating System.

**Comment by pythonmobile** [comment]
I have CUDA 11.1 - is there any chance I can install pytorch using pip on my system correctly?

**Comment by LukeAI** [comment]
@pythonmobile 
it looks like your best bet is
```pip install torch==1.7.0+cu110 torchvision==0.8.1+cu110 torchaudio===0.7.0 -f https://download.pytorch.org/whl/torch_stable.html```

**Comment by stas00** [comment]
pytorch doesn't depend on what cuda you have on your system as it supplies its own.

So what @LukeAI said - it comes from instructions posted at https://pytorch.org/get-started/locally/

**Comment by LukeAI** [comment]
Is it clear now which CUDA 1.7.1 will target?

**Comment by seemethere** [comment]
Yes, unfortunately there wasn't a tenable option to get CUDA 11.1 in for the 1.7.1 release.

We're targeting getting built binaries in for the nightlies and are now looking towards 1.8 for continued CUDA 11.X+ support.

Sorry for the inconvenience, there just wasn't anyway to get CUDA 11.1 in as well as Python 3.9 for 1.7.1

**Comment by stas00** [comment]
cuda-11.2 is officially out! https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html

also available in .deb format - ubuntu, etc.

**Comment by Zhou-Yucheng** [comment]
> @pythonmobile, pytorch doesn't depend on what cuda you have on your system as it supplies its own.
> 
> So what @LukeAI said - it comes from instructions posted at https://pytorch.org/get-started/locally/

Thank you!
So, if I don't get it wrong, I can install CUDA 11.1 or 11.2 and use pytorch1.7.1+cu110?
And in the future, when PyTorch supports CUDA 11.x, I can just upgrade PyTorch via pip and continue to use it?

**Comment by stas00** [comment]
That is correct, just make sure to follow the exact instructions at https://pytorch.org/get-started/locally/

**Comment by hypadr1v3** [comment]
> cuda-11.2 is officially out! https://docs.nvidia.com/cuda/cuda-toolkit-release-notes/index.html
> 
> also available in .deb format - ubuntu, etc.

Great! I got a 3090 2 days agp so was about to go reinstall CUDA and thought that the safest option was just to build pytorch from source. Will CUDA 11.2 be compatible with a CUDNN that is not for the specific version(as 8.0.5 for 11.2 doesn't seem to be out yet)? Also it would be much easier to install a binary instead so would it be just as good as building a source version for my specific card or I may have performance reductions?

**Comment by Rishi5583** [comment]
After installing pytorch lastest in cuda version11.0 , in win 10 , not able to import ,showing some different kind of issues:
I have tried with conda , it fails , then I have installed via pip 
OMP: Error #15: Initializing libiomp5md.dll, but found libiomp5md.dll already initialized.
OMP: Hint This means that multiple copies of the OpenMP runtime have been linked into the program. That is dangerous, since it can degrade performance or cause incorrect results. The best thing to do is to ensure that only a single OpenMP runtime is linked into the process, e.g. by avoiding static linking of the OpenMP runtime in any library. As an unsafe, unsupported, undocumented workaround you can set the environment variable KMP_DUPLICATE_LIB_OK=TRUE to allow the program to continue to execute, but that may cause crashes or silently produce incorrect results. For more information, please see http://www.intel.com/software/products/support/.

**Comment by kaz4ikeda** [comment]
I found a work around (I don't know why): 
Import numpy before importing torch removed OMP error.
I pip-installed pytorch latest in Python 3.8.7, cuda 11.2, Win10 and numpy 1.19.4+mkl.

**Comment by realiti4** [comment]
> I found a work around (I don't know why):
> Import numpy before importing torch removed OMP error.
> I pip-installed pytorch latest in Python 3.8.7, cuda 11.2, Win10 and numpy 1.19.4+mkl.

I think it is a problem with Christoph Gohlke's binaries. I am not sure what changed, but if you delete libiomp dlls in numpys DLLs folder, the errors goes. I am not sure about deleting those though, so your solution works better, but it is a real pain for other libraries you want to use or building a library that using pytorch etc. You need to manually edit files to import numpy first before torch.
